mesonbuild/mcompile.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2020 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Entrypoint script for backend agnostic compile."""
+
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+import typing as T
+
+from . import mlog
+
+if T.TYPE_CHECKING:
+    import argparse
+
+
+def add_arguments(parser: 'argparse.ArgumentParser') -> None:
+    """Add compile specific arguments."""
+    parser.add_argument(
+        '-j', '--jobs',
+        action='store',
+        default=0,
+        type=int,
+        help='The number of worker jobs to run (if supported). If the value is less than 1 the build program will guess.'
+    )
+    parser.add_argument(
+        '-l', '--load-average',
+        action='store',
+        default=0,
+        type=int,
+        help='The system load average to try to maintain (if supported)'
+    )
+    parser.add_argument(
+        '--debug',
+        action='store_true',
+        help='Print debug information for the meson compile command.'
+    )
+    parser.add_argument(
+        '--clean',
+        action='store_true',
+        help='Clean the build directory.'
+    )
+    parser.add_argument(
+        'builddir',
+        type=pathlib.Path,
+        nargs='?',
+        default='.',
+        help='The directory containing build files to be built.'
+    )
+
+
+def _run(options: 'argparse.Namespace') -> int:
+    if options.debug:
+        mlog.log_file = sys.stderr
+        mlog.set_verbose()
+
+    bdir = options.builddir  # type: pathlib.Path
+    if not bdir.exists():
+        mlog.error('Path to builddir {} does not exist!'.format(bdir.as_posix()))
+    if not bdir.is_dir():
+        mlog.error('builddir path should be a directory.')
+
+    cmd = []  # type: T.List[str]
+    runner = None  # type T.Optional[str]
+    if (bdir / 'build.ninja').exists():
+        mlog.debug('Found build.ninja, using ninja.')
+        runner = os.environ.get('NINJA')
+        mlog.debug('Got runner {} from environment variable NINJA'.format(runner))
+        if not runner:
+            if shutil.which('ninja'):
+                runner = 'ninja'
+            elif shutil.which('samu'):
+                runner = 'samu'
+            mlog.debug('found runner {}'.format(runner))
+
+        if runner is None:
+            mlog.error('Cannot find either ninja or samu.')
+            return 1
```

Comment:
Shouldn't we standardize on raising a MesonException in case of error? I had similar questioning for minstall.py tbh, where it's a mix IIRC.